### DISCUSSION TITLE
“SoftwareSerial” PIO-based UART Help

### DISCUSSION THREAD

**Discussion by razaula** [earlephilhower/arduino-pico#1578] [General]
I am wondering if I can inquire for anyone's help on knowing how to implement the Software Serial Emulator correctly for the RP2040 board in Arduino IDE? I am trying to implement the Software Serial Emulator, and have followed the documents in implementing this but unfortunately nothing is outputting onto the serial monitor. I have attached pictures as well. What is a good solution to fix this?

![image](https://github.com/earlephilhower/arduino-pico/assets/59939664/41764deb-5e9a-46ec-937a-8161056381b6)
![image (3)](https://github.com/earlephilhower/arduino-pico/assets/59939664/812b8283-a5d8-4607-b0d7-80027d987a20)

**Comment by earlephilhower** [comment]
You can use the `SoftwareSerial.h` header and object just fine in this core, no need to change to `SerialPIO` (the SWSerial wrapper does that for you).

In any case, the following example outputs data just fine to GP1.  I suggest you check your wiring (remember the PCB pin is not always the same number as the GPx pin):
````
#include <SoftwareSerial.h>

SoftwareSerial s(0,1);

void setup() {
  s.begin(9600);
}

int cnt = 0;
void loop() {
  Serial.printf("send PING: %d\n", cnt);
  s.printf("PING: %d\n", cnt++);
  delay(1000);
}
````

![image](https://github.com/earlephilhower/arduino-pico/assets/11875/c4e11b7d-38e3-4a19-ab87-c2341e20ed42)

**Reply by razaula** [reply]
Thank you for your help. I do see that the bits are getting sent. However, I am confused why s.printf("PING: %d\n",cnt++) is not getting printed to serial monitor. When I tested this on my own, only Serial.printf("PING: %d\n",cnt) was printing with the increase cnt, but s.printf() did not print an individual statement of "PING: %d\n",cnt++

**Reply by earlephilhower** [reply]
The Arduino IDE's serial monitor is connected to the USB port, not any HW or SW UART.  To read UART ports you need a UART<->USB adapter wired up to the UART pins you've defined.

**Reply by razaula** [reply]
Got it. i see thank you very much. My ultimate question is actually knowing how to integrate a raspberry pico with the Sparfun M6E Nano UHF RFID. We want to do a constant read, and after studying the example code (provided in the attachment), I thought it was an issue with the SoftwareSerial library, but now I am thinking that is not it due to discussion. If you have worked with a UHF RDID, the main error we received is "please check wiring". In the code, we are only using functions Serial.begin(115200), Serial.println(), softserial.available(), softserial.isListening(), and softserial.read(), and have connected the wires appropriately. If these functions work as intended in Arduino IDE, is this is mainly an issue with the UHF RFID then?

[RFID_Constant_Read.txt](https://github.com/earlephilhower/arduino-pico/files/11960559/RFID_Constant_Read.txt)
